### PR TITLE
Voltz Vault and LP Optimiser Strategy

### PR DESCRIPTION
**Pull request by Sol3iman**
- Generalisable VoltzVault.sol contract to deposit funds into Voltz pools and accompanying interfaces 
- LPOptimiserStrategy.sol contract which rebalances out of bounds positions based on fixed rates
- Configuration changes to add compatibility with VoltzVault and LPOptimiserStrategy.sol

### CODE REVIEW

**Review by armoking32 — CHANGES_REQUESTED** [review]

File: contracts/interfaces/vaults/IVoltzVault.sol [hunk 1]
```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "./IIntegrationVault.sol";
+import "../external/voltz/IMarginEngine.sol";
+import "../external/voltz/IPeriphery.sol";
+import "../external/voltz/IVAMM.sol";
+import "../external/voltz/rate_oracles/IRateOracle.sol";
```
[the remark this answers is not in the document]
Comment (avniculae) [reply contracts/interfaces/vaults/IVoltzVault.sol:8]: We need this for the getter functions

File: contracts/interfaces/vaults/IVoltzVault.sol [hunk 2]
```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "./IIntegrationVault.sol";
+import "../external/voltz/IMarginEngine.sol";
+import "../external/voltz/IPeriphery.sol";
+import "../external/voltz/IVAMM.sol";
```
[the remark this answers is not in the document]
Comment (avniculae) [reply contracts/interfaces/vaults/IVoltzVault.sol:7]: We need this for the getter functions

File: contracts/interfaces/vaults/IVoltzVault.sol [hunk 3]
```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "./IIntegrationVault.sol";
+import "../external/voltz/IMarginEngine.sol";
```
[the remark this answers is not in the document]
Comment (avniculae) [reply contracts/interfaces/vaults/IVoltzVault.sol:5]: We need this for the getter functions

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        int24 logProximity = _logProximity;
+
+        // 1. Get current position, lower, and upper ticks form VoltzVault.sol
+        IVoltzVault.TickRange memory currentPosition = _vault.currentPosition();
+
+        // 2. Get current tick
+        int24 currentTick = _periphery.getCurrentTick(_marginEngine);
+
+        // 3. Compare current fixed rate to lower and upper bounds
+        int24 logProximity = _logProximity;
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:98]: Remove

**Review by armoking32 — CHANGES_REQUESTED** [review]

File: contracts/interfaces/vaults/IVoltzVault.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (armoking32) [reply contracts/interfaces/vaults/IVoltzVault.sol:8]: here you can leave only

```
import "./IIntegrationVault.sol";
import "../external/voltz/IPeriphery.sol";
```
IMarginEngine, IVAMM, IRateOracle will be imported from IPeriphery

**Review by armoking32 — CHANGES_REQUESTED** [review]

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            (newTick /
+                tickSpacing +
+                ((((newTick % tickSpacing) + tickSpacing) % tickSpacing) >= tickSpacing / 2 ? int24(1) : int24(0))) *
+            tickSpacing;
+    }
+
+    /// @notice Convert a fixed rate to a tick in wad
+    /// @param fixedRateWad The fixed rate to be converted to a tick in wad
+    /// @return int256 The tick in wad
+    function convertFixedRateToTick(int256 fixedRateWad) public view returns (int256) {
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:127]: Change mutability to `pure`

**Review by armoking32 — CHANGES_REQUESTED** [review]

**Review by armoking32 — APPROVED** [review]

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    
+        for (uint256 i = 0; i < _vaults.length; i++) {
+            uint256 vaultWeight = _vaultParams[i].weight;
+
+            if (vaultWeight == 0) {
+                continue;
+            }
+
+            // The share of i-th is vaultWeight / sum(vaultParams.weight)
+            vaultShare[0] = FullMath.mulDiv(balances[0], vaultWeight, totalWeight);
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:254]: Due to rounding here, a situation may arise when not all tokens have been distributed. A possible solution is to send all remaining tokens to the vault with the highest weight.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    }
+
+    // IMMUTABLES
+    address[] public _tokens;
+    IERC20Vault public immutable _erc20Vault;
+
+    // INTERNAL STATE
+    IVoltzVault[] internal _vaults;
+    VaultParams[] internal _vaultParams;
+    uint256 _totalWeight; // sum of all vault weights
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:30]: Add `private` access modifier

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+    /// @notice Attach another voltz vault to the strategy
+    /// @param vault_ The address of the vault
+    /// @param vaultParams_ The paremeters of the new vault
+    function addVault(IVoltzVault vault_, VaultParams memory vaultParams_) external {
+        _requireAdmin();
+        _addVault(vault_, vaultParams_);
+    }
+
+    function _addVault(IVoltzVault vault_, VaultParams memory vaultParams_) internal {
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:98]: Add checks for `vaultParams_`

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+    /// @notice Get the parameters of a vault 
+    /// @param index The index of the vault in _vaults
+    function getVaultParams(uint256 index) public view returns (VaultParams memory) {
+        return _vaultParams[index];
+    }
+
+    /// @notice Set the parameters of a vault
+    /// @param index The index of the vault in _vaults
+    /// @param vaultParams_ The new parameters of the vault
+    function setVaultParams(uint256 index, VaultParams memory vaultParams_) external {
```
Comment (armoking32) [inline contracts/strategies/LPOptimiserStrategy.sol:52]: Add checks for `vaultParams_` structure fields. For example: `MINIMUM_FIXED_RATE <= vaultParams_.maxPossibleLowerBoundWad`.

**Review by armoking32 — CHANGES_REQUESTED** [review]

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    // VAULT PARAMETERS
+    struct VaultParams {
+        int256 sigmaWad; // standard deviation parameter in wad 10^18
+        int256 maxPossibleLowerBoundWad; // Maximum Possible Fixed Rate Lower bounds when initiating a rebalance
+        uint256 proximityWad; // closeness parameter in wad 10^18
+        uint256 weight; // weight parameter that decides how many funds are going to this vault
+    }
+
+    // IMMUTABLES
+    address[] public _tokens;
```
[the remark this answers is not in the document]
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:24]: Renamed. This is of type address[] and the non-value types cannot be immutable.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    IERC20Vault public immutable _erc20Vault;
```
[the remark this answers is not in the document]
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:25]: Renamed.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    function setVaultParams(uint256 index, VaultParams memory vaultParams_) external {
+        _requireAdmin();
+        require(index < _vaults.length, ExceptionsLibrary.INVALID_STATE);
+
+        uint256 previousWeight = _vaultParams[index].weight;
+        _vaultParams[index] = vaultParams_;
+        _totalWeight = (_totalWeight + vaultParams_.weight) - previousWeight;
+    }
+
+    // EVENTS
```
[the remark this answers is not in the document]
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:61]: Done.

File: contracts/strategies/LPOptimiserStrategy.sol [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            _addVault(vaults_[i], vaultParams_[i]);
+        }
+
+        emit StrategyDeployment(erc20vault_, vaults_, vaultParams_, admin_);
+    }
+
+    /// @notice Attach another voltz vault to the strategy
+    /// @param vault_ The address of the vault
+    /// @param vaultParams_ The paremeters of the new vault
+    function addVault(IVoltzVault vault_, VaultParams memory vaultParams_) external {
```
[the remark this answers is not in the document]
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:93]: Done.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 6 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:242]: That's right but the remaining will be pushed in the next deposit and so on.

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 7 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol]: Done

File: contracts/strategies/LPOptimiserStrategy.sol (hunk 9 above)
Comment (0xZenus) [reply contracts/strategies/LPOptimiserStrategy.sol:52]: The parameters are not that constrained and the function is admin-gated.

**Review by armoking32 — APPROVED** [review]

**Pull request closed without merging**